packages/utilities/src/asAsync.tsx

```diff
@@ -0,0 +1,147 @@
+
+import * as React from 'react';
+
+interface IAsAsyncState<TProps> {
+  component?: React.ComponentType<TProps> | undefined;
+  error?: Error;
+  placeholder?: JSX.Element;
+  onLoad?: () => void;
+  onError?: (error: Error) => void;
+}
+
+/**
+ * Represents the surface of a module which has a default export.
+ */
+export interface IModuleWithDefault<T> {
+  default: T;
+}
+
+export interface IAsAsyncProps {
+  /**
+   * A placeholder to render while the component loads.
+   */
+  placeholder?: JSX.Element;
+  /**
+   * A callback to be invoked when the component has loaded and rendered.
+   */
+  onLoad?: () => void;
+  /**
+   * A callback to be invoked when the component has failed to load.
+   */
+  onError?: (error: Error) => void;
+}
+
+/**
+ * Produces a component which internally loads the target component before first mount.
+ * The component passes all props through to the loaded component.
+ *
+ * This overload accepts a module with a default export for the component.
+ */
+export function asAsync<TProps, TOwnProps = IAsAsyncProps>(
+  load: () => Promise<{ default: React.ComponentType<TProps>; }>,
+  mapAsyncProps?: (props: TOwnProps) => IAsAsyncProps
+): React.ComponentType<TProps & TOwnProps>;
+/**
+ * Produces a component which internally loads the target component before first mount.
+ * The component passes all props through to the loaded component.
+ *
+ * This overload accepts a component directly.
+ */
+export function asAsync<TProps, TOwnProps = IAsAsyncProps>(
+  load: () => Promise<React.ComponentType<TProps>>,
+  mapAsyncProps?: (props: TOwnProps) => IAsAsyncProps
+): React.ComponentType<TProps & TOwnProps>;
+/**
+ * Implementatio of asAsync.
+ */
+export function asAsync<TProps, TOwnProps = IAsAsyncProps>(
+  load: () => Promise<React.ComponentType<TProps> | IModuleWithDefault<React.ComponentType<TProps>>>,
+  mapAsyncProps?: (props: TOwnProps) => IAsAsyncProps
+): React.ComponentType<TProps & TOwnProps> {
+  class Async extends React.PureComponent<TProps & TOwnProps, IAsAsyncState<TProps>> {
```

Comment:
I just found that PureComponent is not in 0.14 react, and this ends up breaking it. We have 1 other reference added, and removing it unbreaks react 0.14.

I would like to just remove the peer dependency, but feels like a major change.